src/diagram/workflow_builder.rs

```diff
@@ -0,0 +1,547 @@
+                ordered_outputs.push(o);
+            }
+
+            let joined_output = if join_op.no_serialize.unwrap_or(false) {
+                let join_impl = &registry.join_impls[&ordered_outputs[0].type_id];
+                join_impl(builder, ordered_outputs)?
+            } else {
+                serialize_and_join(builder, registry, ordered_outputs)?.into()
+            };
+
+            let out_edge = edges.get_mut(&target.out_edges[0]).unwrap();
+            out_edge.state = EdgeState::Ready {
+                output: joined_output,
+                origin: None,
+            };
+            Ok(())
+        }
+        // for other operations, each edge is independent, so we can connect at the edge level.
+        _ => {
+            for edge_id in target.in_edges.iter() {
+                connect_edge(builder, registry, edges, inputs, *edge_id, target)?;
+            }
+            Ok(())
+        }
+    }
+}
+
+fn connect_edge<'a>(
+    builder: &mut Builder,
+    registry: &NodeRegistry,
+    edges: &mut HashMap<usize, Edge<'a>>,
+    inputs: &HashMap<&OperationId, DynInputSlot>,
+    edge_id: usize,
+    target: &Vertex,
+) -> Result<(), DiagramError> {
+    let edge = edges.remove(&edge_id).unwrap();
+    debug!(
+        "connect edge {:?}, source: {:?}, target: {:?}",
+        edge_id, edge.source, edge.target
+    );
+    let (output, origin) = match edge.state {
+        EdgeState::Ready {
+            output,
+            origin: origin_node,
+        } => {
+            if let Some(origin_node) = origin_node {
+                (output, Some(origin_node))
+            } else {
+                (output, None)
+            }
+        }
+        EdgeState::Pending => panic!("can only connect ready edges"),
+    };
+
+    match target.op {
+        DiagramOperation::Node(_) => {
+            let input = inputs[target.op_id];
+            let deserialized_output =
+                deserialize(builder, registry, output, target, input.type_id)?;
+            dyn_connect(builder, deserialized_output, input)?;
+        }
+        DiagramOperation::ForkClone(fork_clone_op) => {
+            let amount = fork_clone_op.next.len();
+            let outputs = if output.type_id == TypeId::of::<serde_json::Value>() {
+                <DefaultImpl as DynForkClone<serde_json::Value>>::dyn_fork_clone(
+                    builder, output, amount,
+                )
+            } else {
+                let origin = if let Some(origin_node) = origin {
+                    origin_node
+                } else {
+                    return Err(DiagramError::NotCloneable);
+                };
+
+                let reg = registry.get_registration(&origin.builder)?;
+                reg.fork_clone(builder, output, amount)
+            }?;
+            outputs
+                .into_iter()
+                .zip(target.out_edges.iter())
+                .for_each(|(o, e)| {
+                    let out_edge = edges.get_mut(e).unwrap();
```

Comment:
Safety rationale or `DiagramError`

Also I recommend using conventional for-loops instead of `for_each` if you don't have a very long chain leading up to it. Conventional for-loops make error handling more ergonomic.